### ISSUE TITLE
[Troubleshoot]: Unhandled exception: N5boost10wrapexceptINS_6system12system_errorEEE. what(): thread: Resource temporarily unavailable [system:11]

### ISSUE THREAD

**Issue by Nightmare30** [RLinf/RLinf#359]
### Problem description

Install RLinf by uv for openpi. After build the `Placement`, occur `pthread_create failed: Resource temporarily unavailable`. If use all 8 GPUS, it will failed without entering the train loop.

### Configuration YAML file

YAML config file:

You can also paste the full config here.
```

```

### Log file

You can find the log file in `logs/` folder or the $output_dir/$experiment_name folder (defined in the yaml config) if you are using our example scripts.

Log file:

If you cannot find the log, please provide the full log messages here.
```

[wrong_log.txt](https://github.com/user-attachments/files/23853948/wrong_log.txt)

```

### Environment

python -V:Python 3.11.10
uv pip list: 

**
absl-py                            2.3.1
accelerate                         1.12.0
aiohappyeyeballs                   2.6.1
aiohttp                            3.13.2
aiohttp-cors                       0.8.1
aiosignal                          1.4.0
annotated-types                    0.7.0
antlr4-python3-runtime             4.11.1
anyio                              4.11.0
argcomplete                        3.6.3
arm-pytorch-utilities              0.4.3
array-record                       0.8.3
asttokens                          3.0.1
astunparse                         1.6.3
attrs                              25.4.0
augmax                             0.4.1
av                                 16.0.1
bddl                               3.6.0
beartype                           0.19.0
beautifulsoup4                     4.14.3
blinker                            1.9.0
boto                               2.49.0
boto3                              1.41.5
botocore                           1.41.5
cachetools                         5.5.2
certifi                            2025.11.12
cffi                               2.0.0
cfgv                               3.5.0
charset-normalizer                 3.4.4
chex                               0.1.90
click                              8.3.1
cloudpickle                        3.1.2
cmake                              4.2.0
colorful                           0.6.0a1
contourpy                          1.3.3
crcmod                             1.7
cryptography                       43.0.3
cycler                             0.12.1
dacite                             1.9.2
datasets                           4.4.1
decorator                          5.2.1
deepdiff                           8.6.1
deprecated                         1.3.1
diffusers                          0.35.2
dill                               0.4.0
distlib                            0.4.0
dm-control                         1.0.34
dm-env                             1.6
dm-tree                            0.1.9
docstring-parser                   0.17.0
donfig                             0.8.1.post1
draccus                            0.10.0
easydict                           1.13
einops                             0.8.1
equinox                            0.13.2
etils                              1.13.0
evdev                              1.9.2
executing                          2.2.1
farama-notifications               0.0.4
fast-kinematics                    0.2.2
fasteners                          0.20
fastjsonschema                     2.21.2
filelock                           3.20.0
flask                              3.1.2
flatbuffers                        25.9.23
flax                               0.10.2
fonttools                          4.60.1
frozenlist                         1.8.0
fsspec                             2025.10.0
future                             1.0.0
gast                               0.6.0
gcs-oauth2-boto-plugin             3.3
gcsfs                              2025.10.0
gdown                              5.2.0
gitdb                              4.0.12
gitpython                          3.1.45
glfw                               2.10.0
google-api-core                    2.28.1
google-apitools                    0.5.35
google-auth                        2.39.0
google-auth-httplib2               0.2.1
google-auth-oauthlib               1.2.3
google-cloud-core                  2.5.0
google-cloud-storage               3.6.0
google-crc32c                      1.7.1
google-pasta                       0.2.0
google-reauth                      0.1.1
google-resumable-media             2.8.0
googleapis-common-protos           1.72.0
grpcio                             1.76.0
gsutil                             5.35
gym                                0.26.2
gym-aloha                          0.1.3
gym-notices                        0.1.0
gymnasium                          0.29.1
h11                                0.16.0
h5py                               3.15.1
hf-transfer                        0.1.9
hf-xet                             1.2.1rc0
httpcore                           1.0.9
httplib2                           0.20.4
httpx                              0.28.1
huggingface-hub                    0.36.0
humanize                           4.14.0
hydra-core                         1.4.0.dev1
icmplib                            3.0.4
identify                           2.6.15
idna                               3.11
imageio                            2.37.2
imageio-ffmpeg                     0.6.0
immutabledict                      4.2.2
importlib-metadata                 8.0.0
importlib-resources                6.5.2
iniconfig                          2.3.0
inquirerpy                         0.3.4
ipython                            9.7.0
ipython-pygments-lexers            1.1.1
itsdangerous                       2.2.0
jax                                0.5.3
jax-cuda12-pjrt                    0.5.3
jax-cuda12-plugin                  0.5.3
jaxlib                             0.5.3
jaxtyping                          0.2.36
jedi                               0.19.2
jinja2                             3.1.6
jmespath                           1.0.1
joblib                             1.5.2
jsonlines                          4.0.0
jsonschema                         4.25.1
jsonschema-specifications          2025.9.1
jupyter-core                       5.9.1
jupytext                           1.19.0.dev0
keras                              3.12.0
kiwisolver                         1.4.10rc0
labmaze                            1.0.6
latex2sympy2                       1.9.0
lerobot                            0.1.0
libclang                           18.1.1
libero                             0.1.0
liger-kernel                       0.6.4
llvmlite                           0.44.0
lxml                               6.0.2
mani-skill                         3.0.0b21
markdown                           3.10
markdown-it-py                     4.0.0
markupsafe                         3.0.3
matplotlib                         3.10.7
matplotlib-inline                  0.2.1
mdit-py-plugins                    0.5.0
mdurl                              0.1.2
mergedeep                          1.3.4
ml-collections                     1.0.0
ml-dtypes                          0.5.4
monotonic                          1.6
mplib                              0.1.1
mpmath                             1.3.0
msgpack                            1.1.2
mujoco                             3.3.7
multidict                          6.7.0
multiprocess                       0.70.18
mypy-extensions                    1.1.0
namex                              0.1.0
nbformat                           5.10.4
nest-asyncio                       1.6.0
networkx                           3.6
ninja                              1.13.0
nltk                               3.9.2
nodeenv                            1.9.1
numba                              0.61.2
numcodecs                          0.16.5
numpy                              1.26.4
numpydantic                        1.7.0
nvidia-cublas-cu12                 12.4.5.8
nvidia-cuda-cupti-cu12             12.4.127
nvidia-cuda-nvcc-cu12              12.9.86
nvidia-cuda-nvrtc-cu12             12.4.127
nvidia-cuda-runtime-cu12           12.4.127
nvidia-cudnn-cu12                  9.1.0.70
nvidia-cufft-cu12                  11.2.1.3
nvidia-curand-cu12                 10.3.5.147
nvidia-cusolver-cu12               11.6.1.9
nvidia-cusparse-cu12               12.3.1.170
nvidia-cusparselt-cu12             0.6.2
nvidia-ml-py                       13.580.82
nvidia-nccl-cu12                   2.21.5
nvidia-nvjitlink-cu12              12.4.127
nvidia-nvtx-cu12                   12.4.127
nvitop                             1.6.0
oauth2client                       4.1.3
oauthlib                           3.3.1
omegaconf                          2.4.0.dev4
opencensus                         0.11.4
opencensus-context                 0.2.dev0
opencv-python                      4.11.0.86
opencv-python-headless             4.11.0.86
openexr                            3.4.4
openpi                             0.1.0
openpi-client                      0.1.0
opentelemetry-api                  1.26.0
opentelemetry-exporter-prometheus  0.47b0
opentelemetry-proto                1.38.0
opentelemetry-sdk                  1.26.0
opentelemetry-semantic-conventions 0.47b0
opt-einsum                         3.4.0
optax                              0.2.6
optree                             0.18.0
orbax-checkpoint                   0.11.13
orderly-set                        5.5.0
orjson                             3.11.4
packaging                          25.0
pandas                             2.3.3
parso                              0.8.5
peft                               0.11.1
pexpect                            4.9.0
pfzy                               0.3.4
pillow                             12.0.0
pip                                25.3
platformdirs                       4.5.0
pluggy                             1.6.0
polars                             1.35.2
polars-runtime-32                  1.35.2
pre-commit                         4.5.0
prettytable                        3.17.0
prometheus-client                  0.23.1
promise                            2.3
prompt-toolkit                     3.0.52
propcache                          0.4.1
proto-plus                         1.26.1
protobuf                           6.33.1
psutil                             7.1.3
ptyprocess                         0.7.0
pure-eval                          0.2.3
py-spy                             0.4.1
pyarrow                            22.0.0
pyasn1                             0.6.1
pyasn1-modules                     0.4.2
pybind11                           3.0.1
pycparser                          2.23
pydantic                           2.12.5
pydantic-core                      2.41.5
pyecharts                          2.0.9
pygments                           2.19.2
pylatexenc                         3.0a33
pymunk                             7.2.0
pynput                             1.8.1
pyopengl                           3.1.10
pyopenssl                          24.2.1
pyparsing                          3.3.0b1
pyperclip                          1.11.0
pyrealsense2                       2.56.5.9235
pysocks                            1.7.1
pytest                             9.0.1
python-dateutil                    2.9.0.post0
python-xlib                        0.33
pytorch-kinematics                 0.7.6
pytorch-seed                       0.2.0
pytz                               2025.2
pyu2f                              0.1.5
pyvers                             0.1.0
pyyaml                             6.0.3
pyyaml-include                     1.4.1
pyzmq                              27.1.0
ray                                2.47.1
referencing                        0.37.0
regex                              2025.11.3
requests                           2.32.5
requests-oauthlib                  2.0.0
rerun-sdk                          0.23.1
retry-decorator                    2.0a1
rich                               14.2.0
robosuite                          1.4.1
rpds-py                            0.29.0
rsa                                4.7.2
ruff                               0.14.6
s3transfer                         0.15.0
safetensors                        0.7.0
sapien                             3.0.1
scipy                              1.16.3
sentencepiece                      0.2.1
sentry-sdk                         3.0.0a7
setuptools                         75.8.2
simple-parsing                     0.1.7
simplejson                         3.20.2
six                                1.17.0
smart-open                         7.5.0
smmap                              5.0.2
sniffio                            1.3.1
soupsieve                          2.8
stack-data                         0.6.3
svgwrite                           1.4.3
swanlab                            0.7.2
sympy                              1.13.1
tabulate                           0.9.0
tensorboard                        2.20.0
tensorboard-data-server            0.7.2
tensordict                         0.10.0
tensorflow                         2.20.0
tensorflow-datasets                4.9.9
tensorflow-graphics                2020.5.20
tensorflow-metadata                1.17.2
tensorstore                        0.1.79
termcolor                          3.2.0
timm                               0.9.10
tokenizers                         0.21.4
toml                               0.10.2
toolz                              1.1.0
toppra                             0.6.3
torch                              2.6.0
torch-memory-saver                 0.0.9
torchcodec                         0.8.1
torchdata                          0.11.0
torchvision                        0.21.0
tqdm                               4.67.1
tqdm-loggable                      0.2
traitlets                          5.14.3
transformers                       4.53.2
transforms3d                       0.4.2
tree                               0.2.4
treescope                          0.1.10
trimesh                            4.10.0
triton                             3.2.0
typeguard                          4.4.4
typing-extensions                  4.15.0
typing-inspect                     0.9.0
typing-inspection                  0.4.2
tyro                               0.10.0a6
tzdata                             2025.2
urllib3                            2.5.0
virtualenv                         20.35.4
wadler-lindig                      0.1.7
wandb                              0.23.1rc1
wcwidth                            0.2.14
websockets                         15.0.1
werkzeug                           3.1.3
wheel                              0.45.1
word2number                        1.1
wrapt                              2.0.1
xxhash                             3.6.0
yarl                               1.22.0
zarr                               3.1.5
zipp                               3.23.0
**

nvidia-smi: NVIDIA-SMI 550.163.01             Driver Version: 550.163.01     CUDA Version: 12.4
RLinf version: main
Docker image tag: /

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by andylin-hao** [comment]
Hi @Nightmare30, could you check your system's max thread setting by `cat /proc/sys/kernel/threads-max`?

You can try increase the thread limit by `echo 8251890 > /proc/sys/kernel/threads-max`

**Comment by Nightmare30** [comment]
> Hi [@Nightmare30](https://github.com/Nightmare30), could you check your system's max thread setting by `cat /proc/sys/kernel/threads-max`?
> 
> You can try increase the thread limit by `echo 8251890 > /proc/sys/kernel/threads-max`

The `threads-max` is 8049230. In addition, I added `ulimit -Sn $(ulimit -Hn)` before executing the bash script.

**Comment by Nightmare30** [comment]
> Hi [@Nightmare30](https://github.com/Nightmare30), could you check your system's max thread setting by `cat /proc/sys/kernel/threads-max`?
> 
> You can try increase the thread limit by `echo 8251890 > /proc/sys/kernel/threads-max`

Thank you for your reply. I think I've found the problem. The problem is likely due to a missing GPU library, causing the CPU to be used for rendering, which consumes a significant amount of CPU resources.
I encountered the same problem as #367. I deleted the version limit of metaworld in equirement/openpi.txt to install the openpi successfully, but Vulkan ICD warning still exist. Th output of the `vulkaninfo` is:
```
(RLinf) root@ctmt251103192744s0m-5f889b64b5-j52kh:/home/RLinf# vulkaninfo
ERROR: [Loader Message] Code 0 : libGLX_nvidia.so.0: cannot open shared object file: No such file or directory
Cannot create Vulkan instance.
This problem is often caused by a faulty installation of the Vulkan driver or attempting to use a GPU that does not support Vulkan.
ERROR at ./vulkaninfo/vulkaninfo.h:649:vkCreateInstance failed with ERROR_INCOMPATIBLE_DRIVER
```

When I set:

```
cluster:
  num_nodes: 1
  component_placement:
    env: 0
    rollout: 1
    actor: 2
algorithm:
  group_size: 1
  num_group_envs: 2
```

It can run with `thread: Resource temporarily unavailable [system:11]`, but the utilization rate of GPU0 remained at 0.

**Comment by andylin-hao** [comment]
If you are missing `libGLX_nvidia.so.0`, you'll need to install the NVIDIA rendering library via `apt install libnvidia-gl-<driver-version>`, where `driver-version` is the NVIDIA driver version shown in nvidia-smi like `535`. See here: https://docs.nvidia.com/datacenter/tesla/driver-installation-guide/ubuntu.html#compute-only-headless-and-desktop-only-no-compute-installation

**Comment by andylin-hao** [comment]
If the rendering driver is installed successfully, you can find it at `/usr/lib/x86_64-linux-gnu/libGLX_nvidia.so.0`

**Comment by Nightmare30** [comment]
> If you are missing `libGLX_nvidia.so.0`, you'll need to install the NVIDIA rendering library via `apt install libnvidia-gl-<driver-version>`, where `driver-version` is the NVIDIA driver version shown in nvidia-smi like `535`. See here: https://docs.nvidia.com/datacenter/tesla/driver-installation-guide/ubuntu.html#compute-only-headless-and-desktop-only-no-compute-installation

I installed RLinf by `uv` in a Docker container, so this method may not work for me.

**Comment by andylin-hao** [comment]
Yes, this library can only be installed at the host side. It's crucial that the host side have this library for the `nvidia-container-toolkit` to work properly and enable GPU-accelerated rendering inside container. Are you able to install this at the host side?

**Comment by Nightmare30** [comment]
> Yes, this library can only be installed at the host side. It's crucial that the host side have this library for the `nvidia-container-toolkit` to work properly and enable GPU-accelerated rendering inside container. Are you able to install this at the host side?

Thank you for your help. I'm trying to install and map it into a container at the host side.

**Comment by andylin-hao** [comment]
OK. You may not need to manually map the library into the container. The `nvidia-container-toolkit` can map it automatically for you.

**Comment by Nightmare30** [comment]
> OK. You may not need to manually map the library into the container. The `nvidia-container-toolkit` can map it automatically for you.

Thank you, this solve my problem.

**Comment by Nightmare30** [comment]
> OK. You may not need to manually map the library into the container. The `nvidia-container-toolkit` can map it automatically for you.

Thank you for your help, but I was wrong to attribute the `resource unavailable` to CPU rendering. The real reason is a bug in Ray 2.47.1's resource management. Limiting the number of CPUs from 98 to 16 allows for stable operation without errors.
